Source/Core/Core/HW/DSPHLE/UCodes/ASnd.cpp

```diff
@@ -0,0 +1,544 @@
+// that are not relevant to the DSP side.
+// The old flags are used in the 2008 and 2009 versions, while the new ones are used in the 2011 and
+// 2020 versions.
+constexpr u32 OLD_FLAGS_VOICE_PAUSE = 1 << 5;
+constexpr u32 OLD_FLAGS_VOICE_LOOP = 1 << 2;
+constexpr u32 OLD_FLAGS_SAMPLE_FORMAT_MASK = 3;
+
+constexpr u32 NEW_FLAGS_VOICE_PAUSE = 1 << 9;
+constexpr u32 NEW_FLAGS_VOICE_LOOP = 1 << 8;
+constexpr u32 NEW_FLAGS_SAMPLE_FORMAT_MASK = 7;
+// Used on the PowerPC side as an enabled flag (if 0, then the voice isn't active).
+// Used on the DSP side as the step in bytes after each read.
+constexpr u32 FLAGS_SAMPLE_FORMAT_BYTES_MASK = 0xffff0000;
+constexpr u32 FLAGS_SAMPLE_FORMAT_BYTES_SHIFT = 16;
+
+// November 14, 2008 version (padded to 0x05a0 bytes) - initial release
+// https://github.com/devkitPro/libogc/compare/c76d8b851fafc11b0a5debc0b40842929d5a5825~...353a44f038e75e5982eb550173ec8127ab35e3e3
+constexpr u32 HASH_2008 = 0x8d69a19b;
+// February 5, 2009 version (0x05ac bytes) - added MAIL_TERMINATE
+// https://github.com/devkitPro/libogc/compare/1925217ffb4c97cbee5cf21fa3c0231029b340e2~...3b1f018dbe372859a43bff8560e2525f6efa4433
+constexpr u32 HASH_2009 = 0xc2fd441c;
+// June 11, 2011 version (padded to 0x0620 bytes) - added new sample formats, which shifted flags
+// Note that the source include in the repo does not match the compiled binary exactly; the compiled
+// version differs by using asl instead of lsl, $acc1 instead of $acc0, and $ac0.l instead of $ac0.m
+// in various locations, as well as having the "jmp out_samp" line uncommented in stereo_16bits_le.
+// None of these result in a behavior difference, from the source, though.
+// Note that gcdsptool was also updated, which results in some differences in the source that don't
+// actually correspond to different instructions (e.g. s40 was renamed to s16)
+// https://github.com/devkitPro/libogc/commit/b1b8ecab3af3745c8df0b401abd512bdf5fcc011
+constexpr u32 HASH_2011 = 0xa81582e2;
+// June 12, 2020 version (0x0606 bytes) - libogc switched to compiling the ucode at build time
+// instead of including a pre-compiled version in the repo, so this now corresponds to the code
+// provided in the repo. There appear to be no behavior differences from the 2011 version.
+// https://github.com/devkitPro/libogc/compare/bfb705fe1607a3031d18b65d603975b68a1cffd4~...d20f9bdcfb43260c6c759f4fb98d724931443f93
+constexpr u32 HASH_2020 = 0xdbbeeb61;
+
+constexpr u32 SAMPLE_RATE = 48000;
+
+ASndUCode::ASndUCode(DSPHLE* dsphle, u32 crc) : UCodeInterface(dsphle, crc)
+{
+}
+
+void ASndUCode::Initialize()
+{
+  m_mail_handler.PushMail(DSP_INIT);
+}
+
+void ASndUCode::Update()
+{
+  // This is dubious in general, since we set the interrupt parameter on m_mail_handler.PushMail
+  if (!m_mail_handler.IsEmpty())
+  {
+    DSP::GenerateDSPInterruptFromDSPEmu(DSP::INT_DSP);
+  }
+}
+
+void ASndUCode::HandleMail(u32 mail)
+{
+  if (m_upload_setup_in_progress)
```

Comment:
This if/else seems to operate similarly to how it did in the GBA verson of #10768 before it was refactored.  Intentional?

Reply:
The difference is that there's only two states state (which can be represented by `m_next_command_is_voice_addr`); the ASnd ucode accepts 0xCDD1 commands at all times. The GBA and CARD uCode only accept it once they've finished execution.

Well, there's technically a 3rd state, `m_upload_setup_in_progress`, but that's handled by `UCodeInterface` and it would still exist even if I switched to a 2-entry `State` enum for `ASndUCode`.

... and, on the subject of "state", I noticed that I'm not handling savestates properly here (or in 10768, though it didn't handle them correctly beforehand).  I'll fix that.

Reply:
I've added `DoState` (and also updated 10768 to add it there).